### ISSUE TITLE
bug  计算sft损失的时候

### ISSUE THREAD

**Issue by shyoulala** [GanjinZero/RRHF#48]
计算sft损失的时候label和logits貌似没有shift，是我理解有问题吗？
应该是new_logits = logits[:,:-1,:]

![image](https://github.com/GanjinZero/RRHF/assets/23413415/38e9e92e-a088-494c-8390-e76541b7c45c)

**Comment by shyoulala** [comment]
labels = labels[:,1:,-1]

**Comment by shyoulala** [comment]
这样：
![image](https://github.com/GanjinZero/RRHF/assets/23413415/d4b54cac-0662-4433-abac-6b6e91f25d0b)

**Comment by GanjinZero** [comment]
labels在DataCollatorForSupervisedDataset里shift过了

**Comment by IT-five** [comment]
> DataCollatorForSupervisedDataset
rrhf_loss的loss存在bug,其中`aval = torch.bitwise_and(rw_diff > 0, diff < 0)[0] `，如果我的rw_scores中，最高分排在第一位，由于aval取第0行，则永远到导致aval为空，loss为0，同理scores中，最低分排在第一位，也永远到导致aval为空，为什么偏要取第0行
```
def rrhf_loss(self, scores, idxs, rw_scores):
        diff = scores.unsqueeze(0) - scores.unsqueeze(-1) 
        rw_diff = rw_scores.unsqueeze(0) - rw_scores.unsqueeze(-1) 
        aval = torch.bitwise_and(rw_diff > 0, diff < 0)[0] 
        return -diff[aval].sum() # diff[aval] - shape 
```
